macro/src/check.rs

```diff
@@ -0,0 +1,213 @@
+use crate::{
+    class_info::{ClassInfo, ClassRef, Constructor, Method, RefType, RootMap, Type},
+    reflect::Reflector,
+    span_error::SpanError,
+};
+
+impl RootMap {
+    pub fn check(&self, reflector: &mut Reflector) -> Result<(), SpanError> {
+        let mut errors = vec![];
+
+        for class_name in &self.class_names() {
+            let ci = self.find_class(class_name).unwrap();
+            ci.check(self, reflector, &mut |e| errors.push(e))?;
+        }
+
+        // FIXME: support multiple errors
+        if let Some(e) = errors.pop() {
+            Err(e)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl ClassInfo {
+    fn check(
+        &self,
+        root_map: &RootMap,
+        reflector: &mut Reflector,
+        push_error: &mut impl FnMut(SpanError),
+    ) -> Result<(), SpanError> {
+        let info = reflector.reflect(&self.name, self.span)?;
+
+        let mut push_error_message = |m: String| {
+            push_error(SpanError {
+                span: self.span,
+                message: format!("error in class `{}`: {m}", self.name),
+            });
+        };
+
+        // We always allow people to elide generics, in which case
+        // they are mirroring the "erased" version of the class.
+        //
+        // We need this (at minimum) to deal with `java.lang.Class`, since we
+        // don't want to mirror its parameter.
```

Comment:
Ah, that's actually an interesting problem. I should have added more comments because I'm trying to remember the problem. I think it was this:

If we have `Class<T>`, then we had to include a parameter in the "cached class" methods, and I couldn't figure out what to put there. But you're probably correct that just putting `Class<Object>` would suffice.

It would be nice to always include generic parameters for now (it may be useful to allow them to be erased at some point).

Reply:
I think `Class<Object>` and `Class<?>` have distinct meanings at the Java language level ([reifiable vs non-reifiable types](https://docs.oracle.com/javase/tutorial/java/generics/nonReifiableVarargsType.html)) but are not distinguished at the JVM level.

```java
void test(Class<?> first, Class<? extends Object> second, Class<Object> third, Class fourth) { }
// descriptor: (Ljava/lang/Class;Ljava/lang/Class;Ljava/lang/Class;Ljava/lang/Class;)V
```

Reply:
I guess it also depends on how we want to model the Java generics, since in Java land `Class<String>` is a `Class<?>` but is not a `Class<Object>`.

Reply:
Filed https://github.com/duchess-rs/duchess/issues/41